src/lcancel.c

```diff
@@ -196,35 +197,53 @@ void LCancel_Think(LCancelData *event_data, FighterData *hmn_data)
     JOBJ *hud_jobj = event_data->hud.gobj->hsd_object;
 
     bool should_update_fastfall_hud = false;
+    bool can_fastfall = hmn_data->phys.self_vel.Y < 0;
 
     // log fastfall frame
     // if im in a fastfall-able state
     int state = hmn_data->state_id;
     //if ((state == ASID_JUMPF) || (state == ASID_JUMPB) || (state == ASID_JUMPAERIALF) || (state == ASID_JUMPAERIALB) || (state == ASID_FALL) || (state == ASID_FALLAERIAL) || ((state >= ASID_ATTACKAIRN) && (state <= ASID_ATTACKAIRLW))
     {
-        if (hmn_data->phys.self_vel.Y < 0) // can i fastfall?
+        if (can_fastfall)
         {
             // did i fastfall yet?
             if (hmn_data->flags.is_fastfall)
                 event_data->is_fastfall = 1; // set as fastfall this session
             else
                 event_data->fastfall_frame++; // increment frames
         }
-        else // cant fastfall, reset frames
-        {
-            event_data->fastfall_frame = 0;
-        }
     }
 
-    // if aerial landing
-    if (((hmn_data->state_id >= ASID_LANDINGAIRN) && (hmn_data->state_id <= ASID_LANDINGAIRLW)) && (hmn_data->TM.state_frame == 0))
+    if (is_aerial_landing_state(hmn_data->state_id) && hmn_data->state.frame == 0) {
+        // Record L input timing on the first frame of aerial landing
+        event_data->current_l_input_timing = hmn_data->input.timer_trigger_any_ignore_hitlag;
+    }
+
+    bool has_horizontal_velocity = hmn_data->phys.self_vel.X != 0;
+    bool is_success_l_input = event_data->current_l_input_timing < 7;
+    bool was_previous_state_aerial_landing = is_aerial_landing_state(hmn_data->TM.state_prev[0]);
+
+    bool is_success_l_cancel = is_aerial_landing_state(hmn_data->state_id) && is_success_l_input;
+
+    bool is_missed_l_cancel = is_aerial_landing_state(hmn_data->state_id)
+        && is_success_l_input == false
+        // If there's no horizontal velocity then it's not possible to edge cancel and a miss is confirmed
+        // OR if the number of landing frames has exceeded the number of frames of the l-cancelled animation then it's considered a miss (even if an edge cancel can still occur)
+        && (has_horizontal_velocity == false || ((hmn_data->state.frame + 1) > (hmn_data->figatree_curr->frame_num / 2)));
+
+    bool is_edge_cancel = is_edge_cancel_state(hmn_data->state_id)
+        && was_previous_state_aerial_landing;
+
+    if (event_data->is_current_aerial_counted == false && (is_success_l_cancel || is_missed_l_cancel || is_edge_cancel))
```

Comment:
In C, it's preferable to check for falsity with `!bool_var` instead of `bool_var == false`